# Day 5: Loops & Iteration

---

## Learning Objectives
By the end of this session, you will be able to:
- Write while loops for conditional repetition
- Use for loops with range() for iteration
- Control loop execution with break and continue
- Create nested loops for complex patterns
- Understand loop patterns (counting, summing, searching)
- Avoid infinite loops
- Build practical applications using loops

---

## 1. Introduction to Loops

**Loops** allow you to repeat code multiple times without writing it again and again.

### Why Use Loops?
- Execute code multiple times
- Process sequences of data
- Repeat until a condition is met
- Reduce code duplication

### Types of Loops in Python:
1. **while loop** - Repeats while condition is True
2. **for loop** - Iterates over a sequence

## 2. The while Loop

Repeats as long as a condition is True.

### Syntax:
```python
while condition:
    # code to repeat
    # update condition (important!)
```

### Key Points:
- Condition checked before each iteration
- Must update the condition variable
- Risk of infinite loop if condition never becomes False

In [ ]:
# Simple while loop - countdown
count = 5

while count > 0:
    print(count)
    count -= 1  # IMPORTANT: Update the condition

print("Blastoff! 🚀")

5
4
3
2
1
Blastoff! 🚀


In [ ]:
#DRY RUN
'''
count = 5
count > 0 --- True
print --- 5
count = 5 - 1 = 4
count > 0 --- True
print ----- 4
count = 4 - 1 = 3
count > 0 --- True
print ----- 3
count = 3 - 1 = 2
count > 0 --- True
print ----- 2
count = 2 - 1 = 1
count > 0 --- True
print ----- 1
count = 1
count = 1 - 1 = 0
count > 0 --- False
'''

In [ ]:
# Count from 1 to 10
num = 1

while num <= 10:
    print(num, end=" ")
    num += 1

print("\nDone!")

1 2 3 4 5 6 7 8 9 10 
Done!


In [ ]:
# Calculate sum of first 10 numbers
total = 0
num = 1

while num <= 10:
    total += num
    num += 1

print(f"Sum of 1 to 10: {total}")

Sum of 1 to 10: 55


In [ ]:
'''
DRY RUN
total = 0, num = 1, num <=10 --- True, total = 0 + 1 = 1, num = 1 + 1 = 2
total = 1, num = 2, num <=10 --- True, total = 1 + 2 = 3, num = 2 + 1 = 3
and so on...
'''

### ⚠️ Infinite Loop Warning

An infinite loop runs forever because the condition never becomes False.

```python
# ❌ DANGER: Infinite loop (DON'T RUN!)
# count = 5
# while count > 0:
#     print(count)
#     # Forgot to update count!
```

**How to stop infinite loop:** Press the Stop button or Ctrl+C

### 📝 Exercise 1.1: while Loop Practice

In [ ]:
# TODO: Print even numbers from 2 to 20 using while loop


In [ ]:
# TODO: Calculate factorial of a number using while loop
# factorial of 5 = 5 * 4 * 3 * 2 * 1 = 120
n = int(input("Enter a number: "))
# Your code here


## 3. The for Loop with range()

Used to iterate a specific number of times.

### Syntax:
```python
for variable in range(start, stop, step):
    # code to repeat
```

### range() Function:
- `range(n)` - 0 to n-1
- `range(start, stop)` - start to stop-1
- `range(start, stop, step)` - with custom increment

In [1]:
# Simple for loop: 0 to 4
for i in range(5):
    print(i, end=" ")

0 1 2 3 4 

In [ ]:
'''
range(5) - 0 1 2 3 4
DRY RUN
i = 0, 0
i = 1, 1
i = 2, 2
'''

In [ ]:
# range with start and stop
for i in range(1, 11):
    print(i, end=" ")
print()

1 2 3 4 5 6 7 8 9 10 


In [ ]:
# range with step
print("Even numbers:")
for i in range(0, 21, 2):
    print(i, end=" ")

print("\n\nOdd numbers:")
for i in range(1, 20, 2):
    print(i, end=" ")

Even numbers:
0 2 4 6 8 10 12 14 16 18 20 

Odd numbers:
1 3 5 7 9 11 13 15 17 19 

In [3]:
# Counting backwards
for i in range(100, 49, -1):
    print(i, end=" ")

100 99 98 97 96 95 94 93 92 91 90 89 88 87 86 85 84 83 82 81 80 79 78 77 76 75 74 73 72 71 70 69 68 67 66 65 64 63 62 61 60 59 58 57 56 55 54 53 52 51 50 

### Iterating Over Strings

In [ ]:
'''
i = 'P'
i = 'y'
.
.
.
.

INDEXING IN STRINGS
'P Y T H O N'
 0 1 2 3 4 5
 -6-5-4-3-2-1
'''

In [ ]:
word1 = "Computer 1"
print(len(word1))
print(word1[4])
print(word1[-6])

10
u
u


In [ ]:
# Loop through each character
word = "Python"

for i in word:
    print(i)

#With index
print("\nWith index:")
for i in range(len(word)):
    print(f"Index {i}: {word[i]}")

P
y
t
h
o
n

With index:
Index 0: P
Index 1: y
Index 2: t
Index 3: h
Index 4: o
Index 5: n


### 📝 Exercise 3.1: for Loop Practice

In [ ]:
# TODO: Print multiplication table of 5
# 5 x 1 = 5
# 5 x 2 = 10
# ... up to 5 x 12


In [7]:
number = int(input("Enter the number whose table you want to print: "))
ran = int(input("Enter the range of the table: "))

for i in range(1,ran+1):
  print(f"{number} X {i} = {number*i}")

Enter the number whose table you want to print: 5
Enter the range of the table: 100
5 X 1 = 5
5 X 2 = 10
5 X 3 = 15
5 X 4 = 20
5 X 5 = 25
5 X 6 = 30
5 X 7 = 35
5 X 8 = 40
5 X 9 = 45
5 X 10 = 50
5 X 11 = 55
5 X 12 = 60
5 X 13 = 65
5 X 14 = 70
5 X 15 = 75
5 X 16 = 80
5 X 17 = 85
5 X 18 = 90
5 X 19 = 95
5 X 20 = 100
5 X 21 = 105
5 X 22 = 110
5 X 23 = 115
5 X 24 = 120
5 X 25 = 125
5 X 26 = 130
5 X 27 = 135
5 X 28 = 140
5 X 29 = 145
5 X 30 = 150
5 X 31 = 155
5 X 32 = 160
5 X 33 = 165
5 X 34 = 170
5 X 35 = 175
5 X 36 = 180
5 X 37 = 185
5 X 38 = 190
5 X 39 = 195
5 X 40 = 200
5 X 41 = 205
5 X 42 = 210
5 X 43 = 215
5 X 44 = 220
5 X 45 = 225
5 X 46 = 230
5 X 47 = 235
5 X 48 = 240
5 X 49 = 245
5 X 50 = 250
5 X 51 = 255
5 X 52 = 260
5 X 53 = 265
5 X 54 = 270
5 X 55 = 275
5 X 56 = 280
5 X 57 = 285
5 X 58 = 290
5 X 59 = 295
5 X 60 = 300
5 X 61 = 305
5 X 62 = 310
5 X 63 = 315
5 X 64 = 320
5 X 65 = 325
5 X 66 = 330
5 X 67 = 335
5 X 68 = 340
5 X 69 = 345
5 X 70 = 350
5 X 71 = 355
5 X 72 = 360
5 X 73 = 

In [ ]:
# TODO: Sum all numbers from 1 to 100 using for loop


## 4. Loop Control Statements

Control the flow of loops.

### break Statement
Exits the loop immediately.

In [ ]:
'''
DRY RUN

1 2 3 4 5 6 7 8 9 10
i  = 1, i == 5 --- False, print 1
i  = 2, i == 5 --- False, print 2
.
.
i = 5, i == 5 --- True, break
'''

In [ ]:
# Stop at 5
for i in range(1, 11):
    if i == 5:
        break
    print(i, end=" ")

print("\nLoop stopped at 5")

1 2 3 4 
Loop stopped at 5


In [ ]:
# Find first number divisible by 7
for num in range(1, 100):
    if num % 7 == 0:
        print(f"First number divisible by 7: {num}")
        break

First number divisible by 7: 7


### continue Statement
Skips current iteration, moves to next.

In [ ]:
'''
DRY RUN
i = 1, i % 3 == 0 --- False, print 1
i = 2, i % 3 == 0 --- False, print 2
i = 3, i % 3 == 0 --- True, continue
i = 4, i % 3 == 0 --- False, print 4
.
..
...
'''

In [ ]:
# Skip multiples of 3
for i in range(1, 11):
    if i % 3 == 0:
        continue  # Skip this iteration
    print(i, end=" ")

print("\n(Numbers divisible by 3 were skipped)")

In [ ]:
# Print only positive numbers
numbers = [-5, 3, -2, 8, -1, 10, 0, 7]

print("Positive numbers:")
for num in numbers:
    if num <= 0:
        continue
    print(num, end=" ")

### pass Statement
Does nothing, placeholder for future code.

In [ ]:
# Placeholder loop
for i in range(5):
    if i == 2:
      pass
      # TODO: Add special handling later
    print(i)

0
1
2
3
4


### else Clause with Loops
Executes when loop completes normally (not broken).

In [ ]:
# Search for a number
target = 7

for num in range(1, 11):
    if num == target:
        print(f"Found {target}!")
        break
else:
    print(f"{target} not found in range")

Found 7!


### 📝 Exercise 4.1: Loop Control Practice

In [ ]:
# TODO: Find first 5 prime numbers
# Use break when you have found 5 primes


## 5. Nested Loops

Loops inside loops - useful for patterns, tables, and matrices.

In [ ]:
'''
i = 0
    j = 0, (0,0)
    j = 1, (0,1)
    j = 2, (0,2)
print()
i = 1
    j = 0, (1,0)
    j = 1, (1,1)
    j = 2, (1,2)
print()
i = 2
....
.
.
.
'''

In [ ]:
# Simple nested loop
for i in range(3): # Outer Loop
    for j in range(3): # Inner Loop
        print(f"({i}, {j})", end=" ")
    print()  # New line after inner loop

(0, 0) (0, 1) (0, 2) 
(1, 0) (1, 1) (1, 2) 
(2, 0) (2, 1) (2, 2) 


In [ ]:
"""
i = 1
    j = 1, 1 x 1 = 1  |  j = 2
"""

In [ ]:
# Multiplication table (1 to 5)
print("Multiplication Table:")
print("=" * 30)

for i in range(1, 6):
    for j in range(1, 6):
        print(f"{i} x {j} = {i*j:2d}", end="  |  ")
    print()  # New line

Multiplication Table:
1 x 1 =  1  |  1 x 2 =  2  |  1 x 3 =  3  |  1 x 4 =  4  |  1 x 5 =  5  |  
2 x 1 =  2  |  2 x 2 =  4  |  2 x 3 =  6  |  2 x 4 =  8  |  2 x 5 = 10  |  
3 x 1 =  3  |  3 x 2 =  6  |  3 x 3 =  9  |  3 x 4 = 12  |  3 x 5 = 15  |  
4 x 1 =  4  |  4 x 2 =  8  |  4 x 3 = 12  |  4 x 4 = 16  |  4 x 5 = 20  |  
5 x 1 =  5  |  5 x 2 = 10  |  5 x 3 = 15  |  5 x 4 = 20  |  5 x 5 = 25  |  


### Pattern Printing

In [ ]:
"""
i = 1
    j = 0, *
i = 2
    j = 0, *
    j = 1, *
i = 3
    j = 0, *
    j = 1, *
    j = 2, *
i = 4
    j = 0, *
    j = 1, *
    j = 2, *
    j = 5, *
i = 5
    .
    .
    .
    .
"""

In [ ]:
# Pattern 1: Right triangle
print("Pattern 1: Right Triangle")
for i in range(1, 6):
    for j in range(i):
        print("*", end=" ")
    print()

Pattern 1: Right Triangle
* 
* * 
* * * 
* * * * 
* * * * * 


In [ ]:
# Pattern 2: Inverted triangle
print("\nPattern 2: Inverted Triangle")
for i in range(5, 0, -1):
    for j in range(i):
        print("*", end=" ")
    print()


Pattern 2: Inverted Triangle
* * * * * 
* * * * 
* * * 
* * 
* 


In [ ]:
"""
i = 1
    j  - range(5 - 1) = range(4)
    prints 4 spaces
    k  - range(2*i-1)
    prints * once
i = 2
    j - range(5 - 2) = range(3)
    prints 3 spaces
    k - range(2*i-1)
    prints * twice
i = 3
    .
    .
    .
    .
"""

In [ ]:
# Pattern 3: Pyramid
print("\nPattern 3: Pyramid")
n = 5
for i in range(1, n + 1):
    # Print spaces
    for j in range(n - i):
        print(" ", end="")
    # Print stars
    for k in range(2 * i - 1):
        print("*", end="")
    print()


Pattern 3: Pyramid
    *
   ***
  *****
 *******
*********


In [ ]:
# Pattern 4: Number pyramid
print("\nPattern 4: Number Pyramid")
for i in range(1, 6):
    for j in range(1, i + 1):
        print(j, end=" ")
    print()

### 📝 Exercise 5.1: Pattern Practice

In [ ]:
# TODO: Create a diamond pattern
#     *
#    ***
#   *****
#  *******
# *********
#  *******
#   *****
#    ***
#     *


## 6. Common Loop Patterns

### Pattern 1: Counter

In [ ]:
# Count how many numbers are divisible by 3
count = 0
for num in range(1, 51):
    if num % 3 == 0:
        count += 1

print(f"Numbers divisible by 3 (1-50): {count}")

### Pattern 2: Accumulator (Sum)

In [ ]:
# Sum of squares of first 10 numbers
total = 0
for num in range(1, 11):
    total += num ** 2

print(f"Sum of squares (1-10): {total}")

### Pattern 3: Find Maximum/Minimum

In [ ]:
# Find maximum of user inputs
n = int(input("How many numbers? "))
max_num = float('-inf')  # Start with very small number

for i in range(n):
    num = float(input(f"Enter number {i+1}: "))
    if num > max_num:
        max_num = num

print(f"\nMaximum: {max_num}")

### Pattern 4: Validation Loop

In [ ]:
# Keep asking until valid input
while True:
    age = int(input("Enter age (18-100): "))
    if 18 <= age <= 100:
        print(f"Valid age: {age}")
        break
    else:
        print("Invalid! Try again.")

### Pattern 5: Search

In [ ]:
# Linear search
numbers = [10, 25, 30, 45, 50, 60, 75]
target = int(input("Enter number to search: "))

found = False
position = -1

for i in range(len(numbers)):
    if numbers[i] == target:
        found = True
        position = i
        break

if found:
    print(f"Found {target} at index {position}")
else:
    print(f"{target} not found")

## 7. Practical Applications

### Application 1: Fibonacci Sequence

In [ ]:
# Generate Fibonacci sequence
n = int(input("How many Fibonacci numbers? "))

a, b = 0, 1
count = 0

print("Fibonacci Sequence:")
while count < n:
    print(a, end=" ")
    a, b = b, a + b
    count += 1

How many Fibonacci numbers? 5
Fibonacci Sequence:
0 1 1 2 3 

### Application 2: Prime Number Checker

In [ ]:
# Check if number is prime
num = int(input("Enter a number: "))

if num < 2:
    print(f"{num} is not prime")
else:
    is_prime = True
    for i in range(2, int(num ** 0.5) + 1):
        if num % i == 0:
            is_prime = False
            break

    if is_prime:
        print(f"{num} is prime")
    else:
        print(f"{num} is not prime")

### Application 3: Reverse a Number

In [ ]:
# Reverse digits of a number
num = int(input("Enter a number: "))
original = num
reversed_num = 0

while num > 0:
    digit = num % 10
    reversed_num = reversed_num * 10 + digit
    num = num // 10

print(f"Original: {original}")
print(f"Reversed: {reversed_num}")

# Check palindrome
if original == reversed_num:
    print("It's a palindrome!")

### Application 4: Calculate Digits Sum

In [ ]:
# Sum of digits
num = int(input("Enter a number: "))
original = num
digit_sum = 0

while num > 0:
    digit = num % 10
    digit_sum += digit
    num = num // 10

print(f"Sum of digits of {original}: {digit_sum}")

## 8. Challenge Problems

### 🏆 Challenge 1: Armstrong Number Checker
Armstrong number: sum of cubes of digits equals the number  
Example: 153 = 1³ + 5³ + 3³ = 1 + 125 + 27 = 153

In [ ]:
# TODO: Check if number is Armstrong number
num = int(input("Enter a number: "))
# Your code here


### 🏆 Challenge 2: GCD Calculator
Find Greatest Common Divisor using Euclidean algorithm

In [ ]:
# TODO: Calculate GCD of two numbers
a = int(input("Enter first number: "))
b = int(input("Enter second number: "))
# Your code here


### 🏆 Challenge 3: Perfect Number Finder
Perfect number: sum of proper divisors equals the number  
Example: 6 = 1 + 2 + 3 (divisors of 6)

In [ ]:
# TODO: Check if number is perfect
num = int(input("Enter a number: "))
# Your code here


### 🏆 Challenge 4: Star Pattern - Hourglass

In [ ]:
# TODO: Create hourglass pattern
# *********
#  *******
#   *****
#    ***
#     *
#    ***
#   *****
#  *******
# *********


### 🏆 Challenge 5: Simple ATM System
Create a menu-driven ATM with:
- Check balance
- Deposit
- Withdraw
- Exit
Use a while loop for the menu

In [ ]:
# TODO: ATM System
balance = 10000
# Your code here


## 9. Week 1 Integration Project

### Final Project: Student Grade Management System

Create a complete program that:
1. Takes input for multiple students
2. Calculates grades based on marks
3. Displays statistics (average, highest, lowest)
4. Uses all concepts from Week 1

In [ ]:
# TODO: Complete Student Grade Management System
print("=== Student Grade Management System ===")

num_students = int(input("How many students? "))

# Initialize variables
total_marks = 0
highest = 0
lowest = 100
pass_count = 0

# Your complete solution here
# Use: variables, input, operators, conditionals, loops


## 10. Day 5 & Week 1 Summary

### What We Learned Today (Day 5):

#### Core Concepts:
- ✅ while loops for conditional repetition
- ✅ for loops with range() for counting
- ✅ Loop control: break, continue, pass
- ✅ Nested loops for patterns
- ✅ Common loop patterns (counter, accumulator, search)
- ✅ Avoiding infinite loops

#### Skills Acquired:
- Repeat code efficiently
- Create patterns and tables
- Search and validate data
- Build interactive games
- Solve mathematical problems

---

### 🎉 Week 1 Complete!

#### All Topics Covered:
1. **Day 1:** Python basics, print(), input(), strings
2. **Day 2:** Data types, type conversion, arithmetic operators
3. **Day 3:** Comparison, logical, assignment operators
4. **Day 4:** if-elif-else, nested conditionals
5. **Day 5:** while, for loops, nested loops

#### You Can Now:
- ✅ Write Python programs with proper syntax
- ✅ Work with different data types
- ✅ Perform calculations and comparisons
- ✅ Make decisions with conditionals
- ✅ Repeat code with loops
- ✅ Build complete, interactive programs

---

### Next Week Preview: Data Structures
Week 2 will cover:
- **Day 6:** Lists and Tuples
- **Day 7:** Dictionaries and Sets
- **Day 8:** Functions and Modularity
- **Day 9:** File I/O and Exception Handling
- **Day 10:** Integration Project

---

### 📚 Weekend Practice:
1. Solve all challenge problems
2. Create your own loop-based projects
3. Review all Week 1 concepts
4. Practice pattern printing
5. Build a mini-project combining all Week 1 skills

**Excellent work completing Week 1! 🚀🎓**